src/storage/src/source/mod.rs

```diff
@@ -952,162 +930,191 @@ where
         upstream_name,
         id,
         scope,
-        mut timestamp_histories,
         worker_id,
         worker_count,
         active,
         timestamp_frequency,
         encoding,
+        storage_metadata,
+        as_of,
         base_metrics,
+        now,
         ..
     } = config;
 
     let bytes_read_counter = base_metrics.bytes_read.clone();
 
     let (stream, capability) = source(scope, name.clone(), move |info| {
-        // Create activator for source
-        let activator = scope.activator_for(&info.address[..]);
         let waker_activator = Arc::new(scope.sync_activator_for(&info.address[..]));
         let waker = futures::task::waker(waker_activator);
 
-        // This source will need to be activated when the durability frontier changes.
-        if let Some(wrapper) = timestamp_histories.as_mut() {
-            let durability_activator = scope.activator_for(&info.address[..]);
-            wrapper
-                .wrapper
-                .borrow_mut()
-                .activators
-                .push(durability_activator);
-        }
-
         let metrics_name = upstream_name.unwrap_or_else(|| name.clone());
         let mut source_metrics =
             SourceMetrics::new(base_metrics, &metrics_name, id, &worker_id.to_string());
-        let restored_offsets = timestamp_histories
-            .as_mut()
-            .map(|ts| ts.partitions())
-            .unwrap_or_default();
-        let mut partition_cursors: HashMap<_, _> = restored_offsets
-            .iter()
-            .cloned()
-            .flat_map(|(pid, offset)| Some((pid, offset?)))
-            .collect();
-
-        let mut source_reader = if !active {
-            None
-        } else {
-            match S::new(
+
+        let sync_activator = scope.sync_activator_for(&info.address[..]);
+        let base_metrics = base_metrics.clone();
+        let source_connector = source_connector.clone();
+        let mut source_reader = Box::pin(async_stream::stream! {
+            let mut timestamper = match CreateSourceTimestamper::new(name.clone(), storage_metadata, now, as_of.clone()).await {
+                Ok(t) => t,
+                Err(e) => {
+                    error!("Failed to create source {} timestamper: {}", name, e);
+                    return;
+                }
+            };
+
+            // TODO(petrosagg): What is the purpose of an partition offset being None?
+            //let start_offsets = timestamper.partition_cursors().into_iter().map(|(p, o)| (p, Some(o))).collect();
+            // XXX: Providing these causes some tests to fail: seems like they were NOT being provided before for whatever reason.
+            //      I checked and on main, we pass empty start_offsets in tests like materializations.td
+            let start_offsets = vec![];
```

Comment:
Hm, I looked a bit what happens on `main` and indeed we only ever set the partition offset to `None`. We'll have to persist the rest of the ingestion pipeline in order to be able to skip forward so I think setting it to the empty vec is correct here. I'll keep it in mind for the couple of PRs I have incoming